### ISSUE TITLE
core functions: store()

### ISSUE THREAD

**Issue by cboettig** [cboettig/contenturi#13]
## store

```r
## URL input
store("http://cdiac.ornl.gov/ftp/trends/co2/vostok.icecore.co2")

## local path input
f <- system.file("extdata", "vostok.icecore.co2", package="contenturi", mustWork = TRUE)
store(f)
````

1. Does `store("http://cdiac.ornl.gov/ftp/trends/co2/vostok.icecore.co2")` register both the URL and the local storage location by default? (Yes)
2. Should `store("http://cdiac.ornl.gov/ftp/trends/co2/vostok.icecore.co2")`  have an option to also call to the remote registry (when storing a URL)? (No) 
3. Does `store` allow for the user to indicate which (local) registry &/or store to use?   What does the interface for this look like?  (Currently: `dir = app_dir()` -- you can set the directory used for store+registry, but can't mix & match them)
4. What's the return object? Currently it is the content identifier.  I think that's best for R pipeline chains (particularly useful when a URL is the input), but arguably it should return the local store path instead?

Store always creates a registry entry, but only to a local registry associated with the store.  Thus (unlike the other verbs), store does not have a `registries = ` optional argument.  

@jhpoelen Do you like / agree with the current choices for 1 - 4?  Other issues with `store()`?

**Comment by cboettig** [comment]
Based on feeback in #16: 

- I think `store()` should register any local `source` paths in a separate registry from where it registers URLs.  
- I think the registry format for the local store should follow the `bagit` manifest-sha256.txt format: https://en.wikipedia.org/wiki/BagIt .  (note this would not include date/ timestamp field, though I guess that is arguably much less relevant for local path?)

**Comment by cboettig** [comment]
Rethinking this again.  

- I'm no longer sure `store()` should automatically create any entry in the local registry (i.e. the local `registry.tsv.gz` file that follows the schema discussed).  Registering data in stores is a good idea of course, but the details of this are probably better left to be more explicit, e.g.

```r
store( <some-content-stream> )
register( <a-content-location-of-that-stream> )
```

This reflects the reality that we have lots of different kinds of stores, and stores *just* store (and retrieve) bits & bytes; we don't assume stores know anything about content hash identifiers (or other provenance.) 

Relatedly, it's not entirely clear if this package should even provide implementations of `store()`.  `store()` and `retrieve()` could be left to other packages: the `storr` package provides `st$set()` (i.e. `store`) and `st$get()` (i.e. `retrieve`) for a range of backend "stores".    It might be better if this package is only registry focused, providing `register()` and `query()` functions to interact with said registry but no storage.  

OTOH, having a basic implementation of a local filesystem store is  probably quite useful, and the `bagit_store()` provides a simple mechanism for this.  The current `bagit_store()`, `bagit_retrieve()` could probably be improved a little in the semantics.  We probably also need the ability to `bagit_delete()` objects from a store, and `bagit_list()`, to list all objects in the store (i.e. print the bagit manifest).  This would play nicely with possible higher-level functions like `pin()` (get data specified by a URL location from the local cache or URL, using registries) and `resolve()` (maybe that needs a different verb, but I still want the magic of resolving a content hash to bits & bytes directly, rather than getting a list of possible content locations I need to drum up my own public transit to go fetch!)

**Comment by cboettig** [comment]
@jhpoelen Related to this, would love your thoughts on a minor issue regarding `store()` and streaming content.  

In my dev version, given a URL, `store()` will take file `connections` (including url connections), and it will now stream connections into the store location (using `readBin()` and `writeBin()` rather than using `file.copy()`.  But I'm really not sure this is a good approach!

First, given a URL (`curl`) connection, I cannot just keep it as a URL connection, since that would involve downloading twice: one streaming download to compute the hash, and a second time to stream to disk.  To avoid downloading twice, I download the stream into tempfile, and then stream from tempfile into the storage location (which is based on the hash, like in preston).  

But even so, streaming through `readBin()` and `writeBin()` is much slower than calling `file.copy(temp_location, new_location)` on the paths, and on a practical note I'm not sure what is gained by treating these operation as a connection stream, given that the file is already in temp storage (since we had to hash it) and I don't think there's any user-visible difference (other than the slower speed) for streaming to storage vs `file.copy()`...

download the file to temporary storage before computing a hash id.  It w

**Comment by cboettig** [comment]
Closed by #23
